sstables/trie/bti_index.hh

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
+ */
+
+#pragma once
+
+#include <memory>
+#include "sstables/abstract_index_reader.hh"
+
+namespace sstables {
+    class file_writer;
+    class clustering_info;
+    class deletion_time;
+}
+
+class schema;
+
+namespace sstables::trie {
+
+// IMPORTANT: the index reader is currently page-oriented and
+// it assumes that BTI nodes don't cross page boundaries.
+//
+// Therefore, modifying the page size is a breaking change,
+// because old sstables will violate the assumptions of new readers,
+// and new sstables will violate the assumptions of old readers.
+//
+// If you want to change the page size, you must teach new readers how to handle both
+// the old and the new page size, and new writers must keep writing with the old page size
+// until a cluster feature guarding the new page size is set.
+constexpr static uint64_t BTI_PAGE_SIZE = 4096;
+
+// Transforms a stream of clustering index block entries
+// into a stream of trie nodes fed into bti_trie_sink.
+// Used to populate the Partitions.db file of the BTI format
+// with trie-based indexes of clustering keys.
+class bti_row_index_writer {
+    class impl;
+    std::unique_ptr<impl> _impl;
+public:
+    bti_row_index_writer ();
+    ~bti_row_index_writer ();
+public:
+    // The trie will be written to the given file writer.
+    // Note: the file doesn't have to be empty,
+    // and it can be extended later.
+    bti_row_index_writer(sstables::file_writer&);
+    // Add a new row index entry.
+    // Must be called in ascending order.
+    // (`first_ck` must be strictly greater than the previous `last_ck`).
+    void add(
+        const schema& s,
+        const sstables::clustering_info& first_ck,
+        const sstables::clustering_info& last_ck,
```

Comment:
For the caller, what's the best strategy here? Call every 4k of data? Why not even less?